### PR TITLE
Automatic repair of json for langchains agent

### PR DESCRIPTION
**Pull request by ntrpnr**
Left:  repaired Json
Right: original Json
image

This is a potential fix for #420 #421 #430
Tested an seems to work.

### CODE REVIEW

**Comment by rbren** [comment]
This looks neat! I want to dig into it for a minute but this seems like a great idea

**Comment by enyst** [comment]
I'm not too familiar with the package but IMHO it's worth trying. I did see some other wrong chars occasionally. A few thoughts:
1) we don't want to strip too much and lose track, either 
2) but as long as we log everything, IMHO it's fine
3) I'm going to look into a more extensive logging system. We need it for LLM prompts/responses, stack traces, FE requests, print vars like this.

File: agenthub/langchains_agent/utils/json.py [hunk 1]
```diff
@@ -8,5 +9,11 @@ def dumps(obj, **kwargs):
     return json.dumps(obj, default=my_encoder, **kwargs)
 
 def loads(s, **kwargs):
-    return json.loads(s, **kwargs)
+    s_repaired = repair_json(s)
+
+    if s_repaired != s:
+        print(f"Repaired JSON: {s_repaired}")
+        print(f"Original JSON: {s}")
```
Comment (rbren) [inline agenthub/langchains_agent/utils/json.py:16]: We'll probably want to make these debug logs eventually, but let's leave them in for now so we have some visibility into how this is kicking in for now

**Comment by rbren** [comment]
Took a look at the library. It's definitely a little hacky, but I think it's one of those LLM hacks that probably works very well in practice 😅 I say we get this in

Thanks for the contribution!

**Comment by mangiucugna** [comment]
Hi! Thanks for checking out my library, while the repair actions are a bit hacky, the parser is very solid and at this point it's been tested against a lot of LLMs quite successfully.

One comment here about performance, you will be better off just using the json_repair.loads() drop in replacement instead of the current code, you won't get the diff between original and repaired but it will be faster (in the current code json.loads() and json.dumps() are called a few times).

Then you can run the current code behind a debug flag if you need to check out the diff for debug reasons.

In any case, always happy when someone finds my brainchild useful. If you have any questions or feature requests or bugs please let me know by opening an issue!

Cheers

**Comment by rbren** [comment]
@mangiucugna thanks for the tip! Seems like this will be really helpful for us

**Comment by archenovalis** [comment]
https://github.com/outlines-dev/outlines might be the best solution. when implemented, it will never output improper json, or any chosen schema/regex

**Pull request merged**